### PR TITLE
Fix shell backticks crash

### LINKED ISSUE #13326: Crash when using shell backticks with Bun shell
### How can we reproduce the crash?

```javascript
import { $ } from "bun";
console.log(await $`cd \`find -name dir\``.text());
```

### Relevant log output

```shell
Bun Canary v1.1.22-canary.96 (df33f2b2) macOS Silicon
macOS v14.6.1
Args: "bun" "test.ts"
Features: jsc dotenv shell(2) 
Builtins: "bun:main" 
Elapsed: 55ms | User: 6ms | Sys: 4ms
RSS: 24.28MB | Peak: 24.28MB | Commit: 1.07GB | Faults: 44

panic(main thread): Segmentation fault at address 0xAAAAAAAAAAAAAB22
oh no: Bun has crashed. This indicates a bug in Bun, not your code.
```

### Stack Trace (bun.report)

Bun v1.1.22 ([`df33f2b`](<https://github.com/oven-sh/bun/tree/df33f2b2a264912c7a742d0c03587082415fbba6>)) on macos aarch64 [AutoCommand]

Segmentation fault at address 0xAAAAAAAAAAAAAB22

- *2 unknown/js code*
- `src.shell.interpreter.Interpreter.Cmd.next`
- `src.shell.interpreter.StatePtrUnion`
- `src.shell.interpreter.Interpreter.Expansion.next`
- `src.shell.interpreter.StatePtrUnion`
- `src.shell.interpreter.Interpreter.Expansion.next`
- `src.shell.interpreter.StatePtrUnion`
- `src.shell.interpreter.Interpreter.Expansion.next`
- `src.shell.interpreter.StatePtrUnion`

### PR DESCRIPTION
**Pull request by zackradisic**
### What does this PR do?

Fixes #13326

Add a way to call the IO writing functions safely so they aren't called when they don't need IO

### CODE REVIEW

**Comment by robobun** [comment]
@zackradisic, your commit 5bc6aa22d02b03b467f207c4596d2a2da6fb3a96 is building: [`#2028`](https://buildkite.com/bun/bun/builds/2028)

File: src/shell/interpreter.zig [hunk 1]
```diff
@@ -7098,15 +7135,14 @@ pub const Interpreter = struct {
...
-                    this.bltn.done(0);
+                if (this.bltn.stdout.needsIO()) |safeguard| {
+                    this.state = .waiting;
+                    this.bltn.stdout.enqueue(this, this.output.items[0..], safeguard);
                     return Maybe(void).success;
                 }
-
-                this.state = .waiting;
-                this.bltn.stdout.enqueue(this, this.output.items[0..]);
+                _ = this.bltn.writeNoIO(.stdout, this.output.items[0..]);
```
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig:7143]:
```suggestion
                _ = this.bltn.writeNoIO(.stdout, this.output.items);
```

File: src/shell/interpreter.zig [hunk 2]
```diff
@@ -7335,7 +7379,7 @@ pub const Interpreter = struct {
 
                 switch (errno) {
                     @as(usize, @intFromEnum(bun.C.E.NOTDIR)) => {
-                        if (!this.bltn.stderr.needsIO()) {
+                        if (this.bltn.stderr.needsIO() == null) {
```
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig:7382]: needsIO being an optional seems strange

`needs` implies boolean

File: src/shell/interpreter.zig [hunk 3]
```diff
@@ -5333,26 +5360,35 @@ pub const Interpreter = struct {
...
-                        .fd => true,
-                        else => false,
+                        .fd => OutputNeedsIOSafeGuard{
+                            .__i_know_what_i_am_doing_it_needs_io_yes = 0,
+                        },
+                        else => null,
                     };
                 }
 
+                /// You must check that `.needsIO() == true` before calling this!
```
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig:5372]:
```suggestion
                /// You must check that `.needsIO() != null` before calling this!
```

**Comment by Jarred-Sumner** [comment]
It feels like this should be either a union enum or a regular enum, instead of an optional struct. But I think it's still worth merging

**Pull request merged**